### PR TITLE
Implement Script-Migration

### LINKED ISSUE #1280: Implement Migrations Idempotent script option
The option exists for the k ef migration script command, but the -i flag does not currently have any effect

### LINKED ISSUE #1281: Migrations: Implement Script-Migration ranges
```PowerShell
Script-Migration Version1 Version2
```

Should output a script to take the database from v1 to v2.

### PR DESCRIPTION
**Pull request by bricelam**
Resolves #1280 & resolves #1281

### CODE REVIEW

File: src/EntityFramework.Relational/Migrations/Migrator.cs [hunk 1]
```diff
@@ -166,22 +162,122 @@ public virtual void ApplyMigrations([CanBeNull] string targetMigration = null)
...
+            var builder = new IndentedStringBuilder();
+
+            // If going up...
+            if (string.Compare(fromMigrationName, toMigrationName, StringComparison.OrdinalIgnoreCase) <= 0)
+            {
+                var migrationsToApply = migrations.Where(
+                    m => string.Compare(m.Id, fromMigrationName, StringComparison.OrdinalIgnoreCase) > 0
+                        && string.Compare(m.Id, toMigrationName, StringComparison.OrdinalIgnoreCase) <= 0);
+                var checkFirst = true;
+                foreach (var migraiton in migrationsToApply)
```
Comment (lajones) [inline src/EntityFramework.Relational/Migrations/Migrator.cs]: Typo: migraiton

File: src/EntityFramework.Relational/Migrations/Migrator.cs [hunk 2]
```diff
@@ -166,22 +162,122 @@ public virtual void ApplyMigrations([CanBeNull] string targetMigration = null)
...
+            }
+
+            var builder = new IndentedStringBuilder();
+
+            // If going up...
+            if (string.Compare(fromMigrationName, toMigrationName, StringComparison.OrdinalIgnoreCase) <= 0)
+            {
+                var migrationsToApply = migrations.Where(
+                    m => string.Compare(m.Id, fromMigrationName, StringComparison.OrdinalIgnoreCase) > 0
+                        && string.Compare(m.Id, toMigrationName, StringComparison.OrdinalIgnoreCase) <= 0);
```
Comment (lajones) [inline src/EntityFramework.Relational/Migrations/Migrator.cs]: Are these guaranteed in (ascending) Id order? The going down version below has an OrderByDescending(...).

File: test/EntityFramework.Relational.Tests/Migrations/Infrastructure/MigrationIdGeneratorTest.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        [Fact]
+        public void ResolveId_throws_when_no_match()
+        {
+            var ex = Assert.Throws<InvalidOperationException>(
+                () => new MigrationIdGenerator().ResolveId("Spike", new Migration[0]));
+
+            Assert.Equal(Strings.MigrationNotFound("Spike"), ex.Message);
+        }
+
+        private class MockMigraiton : Migration
```
Comment (lajones) [inline test/EntityFramework.Relational.Tests/Migrations/Infrastructure/MigrationIdGeneratorTest.cs]: Typo - same as above.

**Comment by lajones** [comment]
:shipit:

File: src/EntityFramework.Relational/Migrations/Migrator.cs (hunk 2 above)
Comment (bricelam) [inline src/EntityFramework.Relational/Migrations/Migrator.cs]: Yes, order is ascending by default.

**Pull request merged**